### PR TITLE
Deprecate `_source_include` and `_source_exclude` url parameters

### PR DESCRIPTION
**Pull request by lipsill**
Relates to #22792

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchSourceContext.java [hunk 1]
```diff
@@ -101,23 +105,32 @@ public static FetchSourceContext parseFromRestRequest(RestRequest request) {
...
-                source_includes = Strings.splitStringByCommaToArray(source);
+                sourceIncludes = Strings.splitStringByCommaToArray(source);
             }
         }
+
         String sIncludes = request.param("_source_includes");
-        sIncludes = request.param("_source_include", sIncludes);
+        String sInclude = request.param("_source_include");
+        if (sInclude != null) {
+            DEPRECATION_LOGGER.deprecated("Deprecated field [_source_include] used, expected [_source_includes] instead");
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchSourceContext.java]: What about replacing "field" with "parameter"?

**Review by nik9000 — COMMENTED** [review]
I left something minor but this seems like the right way to do it to me.

Can anyone from @elastic/es-clients comment on the changes to the rest-api-spec?

**Comment by nik9000** [comment]
I labelled this `Search/Search` because I figured that it is *mostly* about the search API. There are a few indices APIs that it effects as well though.

**Comment by nik9000** [comment]
@elasticmachine, test this please

File: rest-api-spec/src/main/resources/rest-api-spec/test/get/70_source_filtering.yml [hunk 2]
```diff
@@ -66,3 +66,37 @@
   - match:   { _id:      "1"       }
   - match:   { fields.count:  [1] }
   - match:   { _source.include.field1: v1 }
+
+---
+"Deprecated _source_include and _source_exclude":
+
+  - skip:
+      version: " - 6.4.99"
```
Comment (lipsill) [inline rest-api-spec/src/main/resources/rest-api-spec/test/get/70_source_filtering.yml]: // TODO consider changing the version (if needed) once the clients specs are changed to `_source_includes` & `_source_excludes`

**Comment by nik9000** [comment]
> @nik9000 thanks for the review! And starting the CI. I wonder why the high level REST client tests passed though. I guess I can look into this next ;)

I'm happy to do it! Why the tests passed is a good question. If they are ignoring deprecation messages that'd be bad but a problem for another PR. Happy hunting!



> Should I assume that this PR will stay on hold until the clients change their specs?

They won't need to change their specs but they might need to do *something*. I'm not sure what. The .NET client generates things from these files.

**Comment by lipsill** [comment]
@nik9000 I just updated the PR to include the the strict deprecation mode and resolve the conflicts.
I was gonna ask somebody to trigger the CI, but it started by itself :innocent:

File: rest-api-spec/src/main/resources/rest-api-spec/test/get/70_source_filtering.yml (hunk 2 above)
Comment (nik9000) [reply rest-api-spec/src/main/resources/rest-api-spec/test/get/70_source_filtering.yml]: I'd use 6.99.99 for now. After I merge this I'll backport it to the 6.x branch. Then I'll come in and update the range here to include the backport target version.

I'm *fairly* sure this is what caused the build to fail. If you crack open the PR build's full console it'll have a `REPRODUCE WITH` line that should reproduce the failure. And I'm fairly sure switching this to 6.99.99 will make it go away.

**Review by nik9000 — CHANGES_REQUESTED** [review]
I left a request to change the line that I *think* caused the build to fail.

**Comment by nik9000** [comment]
Thanks for updating this @lipsill! I think it is super close. Hopefully the `REPRODUCE WITH` line in the build failure will let you reproduce the issue so it isn't too bad to fix. Thanks again!

**Comment by lipsill** [comment]
@nik9000 thanks for the tip work around the issue;)

Can you trigger another CI when you have the chance?

**Comment by nik9000** [comment]
@elasticmachine, test this please

**Review by nik9000 — APPROVED** [review]

**Comment by nik9000** [comment]
@lipsill all merged and backported! Thanks so much for this!

**Comment by Mpdreamz** [comment]
This did end up biting us @nik9000 thank you for the ping though, not sure how it got lost on my end.

**Pull request merged**